framework/sandstone_opts.cpp

```diff
@@ -0,0 +1,848 @@
+    using MaxInteger = std::conditional_t<std::is_signed_v<Integer>, long long, unsigned long long>;
+
+    const char *name = nullptr;
+    const char *explanation = nullptr;
+    MaxInteger min = 0;
+    MaxInteger max = std::numeric_limits<Integer>::max();
+    int base = 10;
+    OutOfRangeMode range_mode = OutOfRangeMode::Exit;
+
+    void print_explanation() const
+    {
+        // i18n style guide says to never construct sentences...
+        if (explanation)
+            fprintf(stderr, "%s: value is %s\n", program_invocation_name, explanation);
+    }
+
+    void print_range_error(const char *arg) const
+    {
+        const char *severity = "warning";
+        if (range_mode == OutOfRangeMode::Exit)
+            severity = "error";
+        if constexpr (std::is_signed_v<Integer>) {
+            fprintf(stderr,
+                    "%s: %s: value out of range for option '%s': %s (minimum is %lld, maximum %lld)\n",
+                    program_invocation_name, severity, name, arg, min, max);
+        } else {
+            fprintf(stderr,
+                    "%s: %s: value out of range for option '%s': %s (minimum is %llu, maximum %llu)\n",
+                    program_invocation_name, severity, name, arg, min, max);
+        }
+        print_explanation();
+    }
+
+    // non-const because this usually comes from optarg anyway
+    Integer operator()(char *arg = optarg) const
+    {
+        assert(name);
+        assert(arg);
+        assert(min <= max);
+        assert(Integer(min) == min);
+        assert(Integer(max) == max);
+
+        char *end = arg;
+        errno = 0;
+        MaxInteger parsed;
+        if constexpr (std::is_signed_v<Integer>)
+            parsed = strtoll(arg, &end, base);
+        else
+            parsed = strtoull(arg, &end, base);
+
+        if (*end != '\0' || *arg == '\0') {
+            // strtoll() did not consume the entire string or there wasn't anything to consume,
+            // so it can't be valid
+            fprintf(stderr, "%s: invalid argument for option '%s': %s\n", program_invocation_name,
+                    name, arg);
+            print_explanation();
+            exit(EX_USAGE);
+        }
+
+        // validate range
+        Integer v = Integer(parsed);
+        bool erange = (errno == ERANGE);
+        if (erange || v != parsed || v < min || v > max) {
+            print_range_error(arg);
+            if (range_mode == OutOfRangeMode::Exit)
+                exit(EX_USAGE);
+
+            if (parsed < min || (erange && parsed == std::numeric_limits<long long>::min()))
+                v = Integer(min);
+            else if (v > max || (erange && parsed == std::numeric_limits<long long>::max()))
+                v = Integer(max);
+        }
+        return v;
+    }
+};
+
+auto parse_testrun_range(const char *arg)
+{
+    int starting_test_number = 1; // One based count for user interface, not zero based
```

Comment:
note that I resigned from passing `starting_test_number` and `ending_test_number` by reference - they are only used in the scope of this function, so that seemed redundant